### DISCUSSION TITLE
How to setup popularimeter for id3v2?

### DISCUSSION THREAD

**Discussion by subframe7536** [Serial-ATA/lofty-rs#401] [Q&A]
Thanks for your great work!

I used to ask for wasm version in https://github.com/Serial-ATA/lofty-rs/discussions/400 and now I am building one myself but dont know how to setup `POPM` frame for id3v2

My code is [here](https://github.com/subframe7536/music-metadata-wasm/blob/16b3c83fd5da9f1e9f17abd504e6120644d08652/src/tag.rs#L117-L126) and I want to achieve something like

```rs
    /// The rate of the song
    #[wasm_bindgen(getter = rate)]
    pub fn get_rate(&self) -> Option<u8> {
        // how?
    }

    #[wasm_bindgen(setter = rate)]
    pub fn set_rate(&mut self, rate: u8) {
        if &self.tag.tag_type() == &TagType::Id3v2 {
            // how?
        } else {
            // how?
        }
    }
```

Any suggestion is highly appreciated!

**Comment by Serial-ATA** [comment]
So `ItemKey::Popularimeter` is very misleading and I actually have it removed for 0.20.0. It only existed to preserve the POPM frame between `Id3v2Tag` -> `Tag` conversions, as all versions before 0.20.0 would have a lossy conversion (#302).

I have been thinking of a way to add a generic `Popularimeter` struct, as it is something that exists in ID3v2, Vorbis Comments, Ilst, and WMA (when that eventually gets supported). For now though, you'll have to create an `Id3v2Tag` to use POPM frames:

```rust
let mut tag = Id3v2Tag::new();

tag.insert(Frame::new(
   "POPM",
   FrameType::Popularimeter(Popularimeter {
      email: String::new(),
      rating: 255,
      counter: 0,
    }),
   FrameFlags::default(),
).unwrap());
```

Note that `Frame` creation will change significantly in 0.20.0.

**Reply by subframe7536** [reply]
so how to achieve this snippet in v0.20.0? 🙏

**Reply by Serial-ATA** [reply]
This should work:

```rust
let tag = Id3v2Tag::new();

tag.insert(Frame::Popularimeter(
  PopularimeterFrame::new(
    String::new(),
    255
    0
  )
));
```

**Reply by subframe7536** [reply]
I'm trying to convert generic `Tag` to `Id3v2Tag` but failed...
```rs
#[wasm_bindgen(setter = rate)]
pub fn set_rate(&mut self, rate: u8) {
    match &self.tag.tag_type() {
        &TagType::Id3v2 => {
            let mut id3v2: &Id3v2Tag = self.tag.into();
            //                                  ^Error
            id3v2.insert(Frame::Popularimeter(PopularimeterFrame::new(
                String::new(),
                rate,
                0,
            )));
        }
        &TagType::VorbisComments => {
            // todo))
        }
        &TagType::Mp4Ilst => {
            // todo))
        }
        &TagType::RiffInfo => {
            // todo))
        }
        _ => {}
    }
}
```
The error info is:
```
the trait bound `&Id3v2Tag: From<Tag>` is not satisfied
the following other types implement trait `From<T>`:
  <Id3v2Tag as From<Tag>>
  <Id3v2Tag as From<v2::tag::SplitTagRemainder>>
required for `Tag` to implement `Into<&Id3v2Tag>`
```

**Reply by Serial-ATA** [reply]
The issue is you're trying to convert to a **&**Id3v2Tag.

Instead of:

```rust
let mut id3v2: &Id3v2Tag = self.tag.into();
```

You'll need:

```rust
let mut id3v2: Id3v2Tag = self.tag.into();
```

**Reply by subframe7536** [reply]
Hmmm... another error prompt: 

cannot move out of `self.tag` which is behind a mutable reference
move occurs because `self.tag` has type `Tag`, which does not implement the `Copy`

I just want to avoid redundant copy

**Reply by Serial-ATA** [reply]
With the way you're using `Tag`, you'll have to do `self.tag.clone()`. `Into::<Id3v2Tag>::into()` consumes the `self.tag` and converts it into an `Id3v2Tag`.

It'd probably be best in your case to wait until 0.21.0 when I make `ItemKey::Popularimeter` available so uou don't have to worry about converting tags.
